pkg/filesystemhelper/filesystemhelper.go

```diff
@@ -178,6 +184,9 @@ func MoveShadow(shadowPath, backupPartsPath string) ([]metadata.Part, int64, err
 		if len(pathParts) != 4 {
 			return nil
 		}
+		if len(partitionsBackupMap) != 0 && !isPartInPartition(pathParts[3], partitionsBackupMap) {
+			return nil
```

Comment:
Wrong algorithm
think about what happens with `shadow` subfolders which is hard link (https://en.wikipedia.org/wiki/Hard_link) and will stay on disk after create backup?

During background merges it will allocate additional disk space after delete original data part

I propose something like `os.Unlink` here, unlink for hard links should be safe and original data parts in `/var/lib/clickhouse/store` should works as expected

Reply:
The data in the shadow subfolders will not disappear, even the merge operation deletes the origin data.
I do not get your idea.

Reply:
Ok. let me try to explain

- you create two data parts in two different partitions `202201_1_1_0` and `202202_1_1_0`
- you create backup only for one partition `202202` during backup you move only one hard link to `/var/lib/clickouse/backup/backup_name` directory
- but, `ALTER TABLE .. FREEZE` create hard links for two data parts
- when you will add one more data part during insert like `202202_2_2_0`, after background merge you will have one data part `202202_1_2_1` `202202_2_2_0` which merged from `202202_1_1_0` and `202202_2_2_0` and data in shadow `202202_1_1_0` which will allocate additional disk space cause original `202202_1_1_0` will delete.

Please read https://en.wikipedia.org/wiki/Hard_link and try to understand how hard link works

Reply:
> Ok. let me try to explain
> 
> * you create two data parts in two different partitions `202201_1_1_0` and `202202_1_1_0`
> * you create backup only for one partition `202202` during backup you move only one hard link to `/var/lib/clickouse/backup/backup_name` directory
> * but, `ALTER TABLE .. FREEZE` create hard links for two data parts
> * when you will add one more data part during insert like `202202_2_2_0`, after background merge you will have one data part `202202_1_2_1` which merged from two parts `202202_1_1_0` and `202202_2_2_0`
>   and data in shadow `202202_1_1_0` which will allocate additional disk space, cause original `202202_1_1_0` will delete.
> 
> Please read https://en.wikipedia.org/wiki/Hard_link and try to understand how hard link works

After freeze the table, we move the parts from shadow  directory to dest backup directory.Then we delete (unlink) all the files in the shadow directory. Then no space is wasted.

Reply:
realSize := map[string]int64{}
disksToPartsMap := map[string][]metadata.Part{}
for _, disk := range diskList {
shadowPath := path.Join(disk.Path, "shadow", backupID)
if _, err := os.Stat(shadowPath); err != nil && os.IsNotExist(err) {
continue
}
backupPath := path.Join(disk.Path, "backup", backupName)
encodedTablePath := path.Join(common.TablePathEncode(table.Database), common.TablePathEncode(table.Name))
backupShadowPath := path.Join(backupPath, "shadow", encodedTablePath, disk.Name)
if err := filesystemhelper.MkdirAll(backupShadowPath, ch); err != nil && !os.IsExist(err) {
return nil, nil, err
}
// If partitionsToBackupMap is not empty, only parts in this partition will be backuped.
parts, size, err := filesystemhelper.MoveShadow(shadowPath, backupShadowPath, partitionsToBackupMap)
if err != nil {
return nil, nil, err
}
realSize[disk.Name] = size
disksToPartsMap[disk.Name] = parts
log.WithField("disk", disk.Name).Debug("shadow moved")
if err := os.RemoveAll(shadowPath); err != nil {
return disksToPartsMap, realSize, err
}
}
if err := filesystemhelper.CleanShadow(backupID, ch); err != nil {
return disksToPartsMap, realSize, err
}


The codes in Bold type clean the shadow directory. 
The RemoveAll will invoke the syscall.Unlink() function, which will delete the files in xx/shadow path.

What's more the code
**if err := filesystemhelper.CleanShadow(backupID, ch); err != nil {
		return disksToPartsMap, realSize, err
	}**

is redundant. We can remove it.

Reply:
great! I glad to know we clarify all aspect of changes